### ISSUE TITLE
Allow formatted strings to all the sources functions that receive a string

### ISSUE THREAD

**Issue by ifraixedes** [bitfield/script#159]
I wondered if all the [source functions](https://pkg.go.dev/github.com/bitfield/script#readme-sources) that receive a string as a parameter could receive a string format and a variadic string parameters as the fmt standard Go package does.

My use case was the following

```
echo \
  "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/debian $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
```

To do that with script you have to execute each subcommand part and do the string interpolation manually.
NOTE this example doesn't illustrate how to handle errors properly to keep it concise.

```go
arch, err := script.Exec("dpkg --print-architecture").String()
if err != nil {
	log.Fatal(err)
}

versionName, err := script.Exec("lsb_release -cs").String()
if err != nil {
	log.Fatal(err)
}

_, err := exec.Echo(fmt.Sprintf("deb [arch=%s signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/debian  %s stable", arch, versionName)).Exec("sudo tee /etc/apt/sources.list.d/docker.list").String()
if err != nil {
	log.Fatal(err)
}
```

If `Echo` would accept the parameters as `fmt.Sprintf` then the last line would be:

```go
_, err := exec.Echo("deb [arch=%s signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/debian  %s stable", arch, versionName).Exec("sudo tee /etc/apt/sources.list.d/docker.list").String()
if err != nil {
	log.Fatal(err)
}
```

It doesn't reduce that much the code, but I thought that it shouldn't hurt and in the end, the variadic strings are optional.

My use case was for only the `Echo` function, but I thought that the other source functions could benefit from the same too.

**Comment by bitfield** [comment]
Yes, this is a useful thing to do, but I'm not sure that `exec(fmt.Sprintf(...))` isn't exactly the right way to do it!

**Comment by ifraixedes** [comment]
I'm sure that from the security point of view is NO unless you can guarantee that the value to interpolate comes from a trusted source.

If I'm doing the `fmt.Sprintf` right now, I'm already committing sins.

Considering that I'm trying to use `script` for what I was writing before in shell script, I don't see any harm. Obviously, the coder is at the end, who has to be aware and not open security wholes.

Do you think that `exec(fmt.Sprintf(...))` isn't exactly the right way to do it because of another point of view ?

**Comment by lazysegtree** [comment]
@bitfield `exec(args)` (fmt.Sprintf done inside exec) is easier to use than `exec(fmt.Sprintf(args))`, and I don't see any obvious security/efficiency problem. 

One drawback I see is that it could unnecessarily bloat the API, if we want to keep separate `Exec`, and `Execf`

**Comment by bitfield** [comment]
A good principle to keep in mind with APIs is Alan Kay's “Simple things should be simple, while complex things should be _possible_”. In other words, make your API so that it elegantly covers 80% of simple use cases without paperwork. Users who need the 20% of more complicated cases won't mind paying the cost of the extra paperwork (in this case, calling `fmt.Sprintf`).

**Comment by lazysegtree** [comment]
That's a nice principle. 👍
